lib/web_ui/lib/src/engine/ulps.dart

```diff
@@ -0,0 +1,194 @@
+// subtraction or between checks.
+//
+// Small introduction into floating point comparison:
+//
+// For some good articles on the topic, see
+// https://randomascii.wordpress.com/category/floating-point/page/2/
+//
+// Here is the 32 bit IEEE representation:
+//   uint32_t mantissa : 23;
+//   uint32_t exponent : 8;
+//   uint32_t sign : 1;
+// As you can see it was carefully designed to be reinterpreted as an integer.
+//
+// Ulps stands for unit in the last place. ulp(x) is the gap between two
+// floating point numbers nearest x.
+
+/// Converts a sign-bit int (float interpreted as int) into a 2s complement
+/// int. Also converts 0x80000000 to 0. Allows result to be compared using
+/// int comparison.
+int signBitTo2sCompliment(int x) => (x & 0x80000000) != 0 ? (-(x & 0x7fffffff)) : x;
+
+/// Convert a 2s compliment int to a sign-bit (i.e. int interpreted as float).
+int twosComplimentToSignBit(int x) {
+  if ((x & 0x80000000) == 0) {
+    return x;
+  }
+  x = ~x + 1;
+  x |= 0x80000000;
+  return x;
+}
+
+class _FloatBitConverter {
+  final Float32List float32List = Float32List(1);
+  late Int32List int32List;
+  _FloatBitConverter() {
+    int32List = float32List.buffer.asInt32List(0, 1);
+  }
+  int toInt(Float32List source, int index) {
+    float32List[0] = source[index];
+    return int32List[0];
+  }
+  int toBits(double x) {
+    float32List[0] = x;
+    return int32List[0];
+  }
+  double toFloat(int bits) {
+    int32List[0] = bits;
+    return float32List[0];
+  }
+}
+
+// Singleton bit converter to prevent typed array allocations.
+_FloatBitConverter _floatBitConverter = _FloatBitConverter();
+
+// Converts float to bits.
+int float2Bits(Float32List source, int index) {
+  return _floatBitConverter.toInt(source, index);
+}
+
+// Converts bits to float.
+double bitsToFloat(int bits) {
+  return _floatBitConverter.toFloat(bits);
+}
+
+const int floatBitsExponentMask = 0x7F800000;
+const int floatBitsMatissaMask  = 0x007FFFFF;
+
+/// Returns a float as 2s compliment int to be able to compare floats to each
+/// other.
+int floatFromListAs2sCompliment(Float32List source, int index) =>
+    signBitTo2sCompliment(float2Bits(source, index));
+
+int floatAs2sCompliment(double x) =>
+    signBitTo2sCompliment(_floatBitConverter.toBits(x));
+
+double twosComplimentAsFloat(int x) =>
+    bitsToFloat(twosComplimentToSignBit(x));
+
+bool _argumentsDenormalized(double a, double b, int epsilon) {
+  double denormalizedCheck = kFltEpsilon * epsilon / 2;
+  return a.abs() <= denormalizedCheck && b.abs() <= denormalizedCheck;
+}
+
+bool equalUlps(double a, double b, int epsilon, int depsilon) {
+  if (_argumentsDenormalized(a, b, depsilon)) {
+    return true;
+  }
+  int aBits = floatAs2sCompliment(a);
+  int bBits = floatAs2sCompliment(b);
+  // Find the difference in ULPs.
+  return aBits < bBits + epsilon && bBits < aBits + epsilon;
+}
+
+bool almostEqualUlps(double a, double b) {
```

Comment:
Docs? (applies here and other undocumented functions)

Reply:
Done.